### PR TITLE
Add documentation site

### LINKED ISSUE #849: Proposal: Documentation for client_python
_(I initially drafted this in Discussions, but I ended moving here. Happy to move it there if required)_

# Context

After digging a bit in this project and reading some of the issues, I found some things that could be improved regarding the documentation:

- The whole API and usage examples live in a single `README.md` making harder to read.
- Some classes and methods are documented but not rendered in any documentation site. Someone also #848 that current docstrings don't follow any standard in order to be rendered properly.
- Documentation is incomplete and could be extended. See issues #499 #528 

# Proposals

## Documentation language

Current README uses markdown to render the contet. The obvious choice would be use markdown, but I just wanted to mention that rewriting to RST (RestructuredText) it's also an option. 

## Documentation Generator

There are many options to generate documentation but I'm only listing here the most suitable options for Python.

### [mkdocs](https://www.mkdocs.org/)

This the default choice for many projects. This [tutorial](https://realpython.com/python-project-documentation-with-mkdocs) explains very well how to use with Python.

Pros:
- Supports Markdown
- Supports [Google-Style Docstrings](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings), [NumPy Docstring Standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) and [Sphinx Docstring Format](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format).
- Has many plugins and lot of community around it.

### [Sphinx](https://www.sphinx-doc.org/en/master/index.html)

This is the classic way to host Python docs, typicall using RST and hosted in readthedocs.

Pros:
- Supports RST and Markdown (with a plugin).
- Classic player for Open Source Python projects.

### [nanoc](https://nanoc.app/)

This is what the official Prometheus documentation uses to generate its documentation.

Pros:
- It documentation system used by Prometheus.

Cons:
- It doesn't generate documentation from Python docstrings.
- Community and adoption is lower compared to other projects.

## Hosting

Where the documentation should live is also relevant for this discussion.

### [GitHub Pages](https://pages.github.com/)

Is it the [recommended](https://realpython.com/python-project-documentation-with-mkdocs/#step-6-host-your-documentation-on-github) way when using mkdocs. It works creating a branch/repo that will publish automatically changes when new docs are pushed to main.

### [Read the Docs](https://readthedocs.org)

Is it a very common way to host documentation pages of Python projects. It works configuring a hook in repository. Readthedocs will pull the changes and publish to the configured site.

### prometheus.io

This approach requires using [prometheus docs](https://github.com/prometheus/docs) along with `nanoc` to host the documentation. It would make everything consistent as all the docs are in the same website and it would open the door to other client libraries to host documentation there. The repo could pull changes from `client_python/docs`. However this would require manually update the [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) everytime docs are changed in `client_python`.

Alternatively, we could create a subdomain (https://prometheus.io/docs/python) that points to a different hosting that the Prometheus project.

## Do nothing

Making the README bigger every time that new features are add and leave docs and classes undocumented.

### PR DESCRIPTION
**Pull request by marctc**
- Added initial structure to render docs with github pages. We decided to use [hugo](https://gohugo.io/) as is the tool used for Prometheus' client_java (see docs [here](https://prometheus.github.io/client_java/)). It's also quite popular tool among open source projects.
- README file is being cleaned up to have minimal information. The content is moved to different subsections. I mostly followed the existing hierarchy but I also moved things a bit around. Please take a look and feel free to suggest how to structure it.
- Added github action to deploy changes when master is built. This probably needs to enable the github page in project settings (@fstab can support us on this in case we can't figure how to do it).

In a follow-up PR, I can add Python docstrings of the documented classes. But for the scope of this PR I decided to leave it out.

Fixes #849 

This a preview how it renders:

### CODE REVIEW

File: .github/workflows/github-pages.yaml
Comment (csmarchbanks) [inline .github/workflows/github-pages.yaml:4]: This comment can be removed since we don't have a 1.0.x branch.

File: docs/content/instrumenting/extra-options.md
Comment (csmarchbanks) [inline docs/content/instrumenting/extra-options.md:6]: It's probably worth having Labels as a separate item since they are a core part of using this library.

**Review by csmarchbanks — COMMENTED** [review]
Just a couple comments, but this looks great!

Is there a good way to deploy this to a github site just to check that CI works before merging? Locally everything looked nice.

File: docs/content/instrumenting/extra-options.md
Comment (marctc) [reply docs/content/instrumenting/extra-options.md:6]: Moved labels and exemplars to their own section. `## Disabling `_created` metrics` is now in the `_index.md` of instrumenting. I can change if need it.

**Comment by marctc** [comment]
> Is there a good way to deploy this to a github site just to check that CI works before merging? 

If I'm not mistaken, the only way to deploy to pages is syncing with a github branch. @fstab did you try to manually deploy the docs? or you just used the github action directly?

**Comment by fstab** [comment]
I just used the Github action. I didn't even write it, I just copied it out of the Hugo documentation :) https://gohugo.io/hosting-and-deployment/hosting-on-github/

**Comment by csmarchbanks** [comment]
There were some readme updates that came in with allowing https, would you be willing to rebase this with those new changes?

Otherwise I think this looks good and I can work on any deployment issues in followup PRs if necessary.

File: docs/content/exporting/http/_index.md
Comment (csmarchbanks) [inline docs/content/exporting/http/_index.md:2]: Perhaps the title here should be HTTP/HTTPS now to make it more clear that it covers both?

File: docs/content/exporting/http/_index.md
Comment (csmarchbanks) [inline docs/content/exporting/http/_index.md:25]: It would be nice to have an HTTPS section header above this.

**Review by csmarchbanks — COMMENTED** [review]
A couple tiny comments, plus the most recent commit needs a DCO signoff.

**Review by csmarchbanks — APPROVED** [review]
Thanks!

**Pull request merged**
